test/router-options.test.js

```diff
@@ -33,3 +33,28 @@ test('honors ignoreTrailingSlash option', t => {
     })
   })
 })
+
+test('Should honor maxParamLength option', t => {
+  t.plan(4)
+  const fastify = Fastify({ maxParamLength: 10 })
+
+  fastify.get('/test/:id', (req, reply) => {
+    reply.send({ hello: 'world' })
+  })
+
+  fastify.inject({
+    method: 'GET',
+    url: '/test/123456789'
+  }, (error, res) => {
+    t.error(error)
+    t.strictEqual(res.statusCode, 200)
+  })
+
+  fastify.inject({
+    method: 'GET',
+    url: '/test/123456789abcd'
+  }, (error, res) => {
+    t.error(error)
+    t.strictEqual(res.statusCode, 404)
```

Comment:
I expect `414 URI Too Long` https://tools.ietf.org/html/rfc7231#section-6.5.12

Reply:
We can discuss a lot about this, is it a 414 or a 400?
The thing is not that the uri is too long, but a single parameter breaks our maximum length.
In any case this check is done internally by `find-my-way`, and is not directly exposed to Fastify unless you use the defaultRoute.

Reply:
The maximum length of an uri is directly handled by Node itself if I'm not wrong.

Reply:
I would agree with `414` or `400` but not for `404` because the route is definitely matched. According to debugging it is really hard we don't log anything.

> In any case this check is done internally by find-my-way, and is not directly exposed to Fastify unless you use the defaultRoute.

This is wrong Fastify is responsible to respond in the correct way with the HTTP spec. The "content validation" is out of the scope of a router IMO.

Reply:
> The maximum length of an uri is directly handled by Node itself if I'm not wrong.

Reply:
> The thing is not that the uri is too long, but a single parameter breaks our maximum length.

Reply:
> The thing is not that the uri is too long, but a single parameter breaks our maximum length.

What's the maximum?

As you said this done internally by node.

Reply:
Again we can discuss about this, is a matter of philosophy :P
The thing is that the limit is there for security reasons, it is pretty hard that you will have ≥ 100 chars parameters. And you can read it as "there are no routes that can handle your request" or "your url is wrong", both cases are true.
Talking from a performances perspective, this implementation is way faster, doing what you suggest means change the core algorithm of `find-my-way` and update also how Fastify handle errors.

IMHO we should keep this as is and if is a problem in the future increase the maximum allowed length or update the algorithm.

Reply:
> What's the maximum?

Read the docs, 100 chars.

> As you said this done internally by node or do I miss something?

The single parameter check is done by `find-my-way`, the overall length of the url is checked by node.

Reply:
> Again we can discuss about this, is a matter of philosophy :P

No chance, `404` would mean that the route could not be matched. We have better alternatives like `400` or `414`. You combine content validation and matching in one category. I don't like it.

> The thing is that the limit is there for security reasons, it is pretty hard that you will have ≥ 100 chars parameters

Why do we check this in general when it's very uncommon and node core already check this security flaw for us?

Reply:
> No chance, 404 would mean that the route could not be matched. We have better alternatives like 400 or 414. You combine content validation and matching in one category. I don't like it.

I understand your point, but I'd like to reming you at my previous answer:
*Talking from a performances perspective, this implementation is way faster, doing what you suggest means change the core algorithm of find-my-way and update also how Fastify handle errors.*

*IMHO we should keep this as is and if is a problem in the future increase the maximum allowed length or update the algorithm.*

> Why do we check this in general when it's very uncommon and node core already check this security flaw for us?

Node checks the overall length of the url.
With this we check the length of a single parameter (eg: `/test/:id`, here we are checking the id length). For example: if you have some kind of regex route and an attacker knows it you will be vulnerable against a DoS attack.

Reply:
> For example: if you have some kind of regex route and an attacker knows it you will be vulnerable against a DoS attack.

```
test/:bigRegex
```
Why are you vulnerable against a DoS attack?

```
Node.js incoming message -> 414 (by node core) -> find-my-way (id validation)
```
A too long URI does not arrive `find-my-way` it will be caught by [node core](https://github.com/nodejs/node/blob/db1087c9757c31a82c50a1eba368d8cba95b57d0/deps/http_parser/http_parser.c#L143).

Reply:
A too long uri is way larger than 100 chars ([this](https://github.com/nodejs/node/blob/master/deps/http_parser/http_parser.h#L56-L65) is the limit), while run a regex on a 1000 parameter could cause a DoS attack. Don't forget that node can perform one operation at the time.

Reply:
yep, but this is not the job of a router. We should fix this some levels before and in fastify core or **provide a plugin**.

Reply:
You can't put it in Fastify core or in a plugin, because the attack will happen only inside the router.

Reply:
"The attack" is filtered by node core at a URI length of `80KB` that's fine. You are talking about a super sensitive solution.

Reply:
https://github.com/fastify/fastify/pull/716#issuecomment-361073045

Reply:
> "The attack" is filtered by node core at a URI length of 80KB that's fine. You are talking about a super sensitive solution.

It is not.
As I said it in https://github.com/fastify/fastify/pull/716#issuecomment-361073030, the amount of data to put a regexp to an endless loop is way smaller, if a complex enough regexp is used.
My _preference_ has always been to not support regexp at all, but that is a very much needed feature of a web framework.

Reply:
Yes, at this point I hadn't understood it fully.